### DISCUSSION TITLE
Feature Request: Add code execution tool for accurate Cyrillic/multilingual text rendering in generated images

### DISCUSSION THREAD

**Discussion by schiller20dima-sketch** [QwenLM/Qwen3.6#228] [Ideas]
## 💡 The Problem

When using Qwen's image generation to create collages or images with text overlays, **Cyrillic (Russian) text comes out heavily garbled**.

Examples of distorted output:
- `чёрнылй матоьоть` instead of `чёрный матовый`
- `серуй прозрачниь` instead of `серый прозрачный`
- `черепаховылий (hanana)` instead of `черепаховый (havana)`

English labels in the same image render correctly, which confirms the issue is specific to non-Latin scripts.

## 🔍 Root Cause

The current image generator **"draws" letters as visual patterns** (pixel by pixel) instead of rendering them with a proper font. Since training data contains far more English text-on-images than Cyrillic, the model fails on Russian glyphs.

## ✅ Proposed Solution

Add a **code execution tool (Python sandbox)** to Qwen's assistant toolset — similar to ChatGPT's Code Interpreter. This would allow:

1. Programmatic image composition using **PIL/Pillow**
2. Accurate text rendering with real **TTF fonts** (DejaVu Sans, Roboto, etc.)
3. **Perfect spelling in any language** — Russian, Arabic, Chinese, etc.

## 🎁 Alternative Solutions

- Fine-tune the image generator on multilingual text-in-image datasets
- Provide a "text-free" image generation mode + structured text output for manual captioning

## 🌍 Why This Matters

Russian has ~258 million speakers worldwide. Accurate multilingual text rendering would dramatically improve Qwen for marketing materials, educational infographics, product mockups, and social media content.

## 📎 Reproduction Steps

1. Upload a personal photo
2. Ask: *"Create a 2x2 collage of me trying on 4 eyeglass frames with Russian labels"*
3. Observe: faces and glasses render perfectly, but Russian captions are gibberish

Thank you for building Qwen — looking forward to future improvements! 🙏
<img width="1664" height="928" alt="1dc8a51ef-5120-48d7-993c-a1b94b16042e" src="https://github.com/user-attachments/assets/f081af0a-febb-4530-a6b8-1d6d3fe957ce" />
